Plutarch/DataRepr/Internal.hs

```diff
@@ -180,7 +184,55 @@ type family PUnLabel (a :: PLabeledType) :: PType where
  This is how most data structures are stored on-chain.
 -}
 type PDataSum :: [[PLabeledType]] -> PType
-data PDataSum (defs :: [[PLabeledType]]) (s :: S)
+data PDataSum (defs :: [[PLabeledType]]) (s :: S) where 
+  DSZ :: (forall {out :: PType}. (Term s (PDataRecord x) -> Term s out) -> Term s out) 
+      -> PDataSum (x ': xs) s
+
+  DSS :: PDataSum xs s -> PDataSum (x ': xs) s
+
+-- the length functions in sop-core gives the actual length, we want length-1
+adjustedLength :: forall x xs proxy. SListI (x:xs) => proxy (x:xs) -> Integer 
+adjustedLength proxy = fromIntegral (lengthSList proxy) - 1  
+
+instance SListI (def:defs) => PlutusType (PDataSum (def:defs)) where
```

Comment:
You're not familiar with its purpose in general, or here? 

In general, it's more or less the generics-sop version of `SingI` from singletons except specialized to lists (and way more annoying to work with). It's basically just a constraint that asserts that the typelevel list is known at compile time. 

Here, we need it for two things: 

1) It's necessary to calculate the length of the typelevel list so we can get the index of the correct variant. This is a consequence of the NS-like representation; we can't just count the constructors because, with the NS-like representation, we don't have a constructor for every element of the typelevel list. (This is why I thought it was initially impossible to use an NS-like representation - I forgot SListI existed. Under the old representation, you *could* count the constructors.) 

2) It's necessary to implement the helper function `go` in `pmatch'`. We need some kind of GADT from which we can recover information about the structure of the typelevel list. I don't think there's any way to make this typecheck without that while rebuilding the sum. I also don't think there's way to pass that information explicitly in the arguments to `pmatch'`, so we have to conjure it into existence, which requites a type class constraint. 

It looks like there *might have been some machinery at one point relating to `HList`s in the Plutarch.DataRepr.Internal.HList / Plutarch.DataRepr.Internal.HList.Utils modules that we could use instead of SListI, but either the actual HList stuff never got added or was replaced with HRec.  

I think we need something roughly equivalent to SListI/SList. Since we're already using Generics.SOP anyway, it seemed like the most parsimonious choice. If you have an idea of how to get around needing the constraint I'd be happy to implement it, but I can't think of anything.